src/librustc_mir/transform/uninhabited_enum_branching.rs

```diff
@@ -0,0 +1,124 @@
+//! A pass that eliminates branches on uninhabited enum variants.
+
+use crate::transform::{MirPass, MirSource};
+use rustc::mir::{
+    BasicBlock, Body, Local, Operand, Place, PlaceBase, Rvalue, StatementKind, TerminatorKind,
+};
+use rustc::ty::layout::{Abi, TyLayout, Variants};
+use rustc::ty::{Ty, TyCtxt};
+
+pub struct UninhabitedEnumBranching;
+
+fn get_discriminant_local(terminator: &TerminatorKind<'_>) -> Option<Local> {
+    if let TerminatorKind::SwitchInt {
+        discr: Operand::Move(Place { base: PlaceBase::Local(l), projection: box [] }),
+        ..
+    } = terminator
+    {
+        Some(*l)
+    } else {
+        None
+    }
+}
+
+/// Finds all basic blocks which terminate by switching on a discriminant.
+/// Returns a `Vec` of block ids and the `Ty` the discriminant is read from.
+fn find_eligable_blocks<'tcx>(body: &Body<'tcx>) -> Vec<(BasicBlock, Ty<'tcx>)> {
```

Comment:
Yeah, I can do that. I think it might also be beneficial to do a check here and make sure that the `TyKind` is an `Adt`. I'm guessing a large number of `switchInt` terminators probably operate on `bool` or `int` values and we can filter those out here before doing any other work.

Reply:
Actually, I'm running into lifetime issues if I try that:

```
error[E0597]: `body` does not live long enough
  --> src/librustc_mir/transform/uninhabited_enum_branching.rs:37:38
   |
22 | fn find_eligible_blocks<'tcx>(body: &Body<'tcx>) -> impl Iterator<Item = (BasicBlock, Ty<'tcx>)> {
   |                         ----                        -------------------------------------------- opaque type requires that `body` is borrowed for `'tcx`
   |                         |
   |                         lifetime `'tcx` defined here
...
26 |         .filter_map(|(bb, block_data)| {
   |                     ------------------ value captured here
...
37 |                             let ty = body.local_decls[r_local].ty;
   |                                      ^^^^ borrowed value does not live long enough
...
49 | }
   | - `body` dropped here while still borrowed

```

Reply:
You need to add an explicit lifetime to the reference and the impl trait:

```rust

fn find_eligible_blocks2<'a, 'tcx>(body: &'a Body<'tcx>) -> impl Iterator<Item = (BasicBlock, Ty<'tcx>)> + 'a {
    body.basic_blocks()
        .iter_enumerated()
        .filter_map(move |(bb, block_data)| {
            let terminator = block_data.terminator();

            // Only bother checking blocks which terminate by switching on a local.
            if let Some(local) = get_discriminant_local(&terminator.kind) {
                let stmt_before_term = (block_data.statements.len() > 0)
                    .then_with(|| &block_data.statements[block_data.statements.len() - 1].kind);

                if let Some(StatementKind::Assign(box (l, Rvalue::Discriminant(place)))) =
                    stmt_before_term
                {
                    if l.as_local() == Some(local) {
                        if let Some(r_local) = place.as_local() {
                            let ty = body.local_decls[r_local].ty;

                            if ty.is_enum() {
                                return Some((bb, ty));
                            }
                        }
                    }
                }
            }
            None
        })
}
```

Also note the `move` on the closure

Reply:
So the other issue is that we borrow `body` here

https://github.com/rust-lang/rust/pull/65694/files#diff-aee445fc8623471c31a62729dbf11e6eR75

so it's borrowed for the lifetime of `blocks_to_update` but then we also try to use it here

https://github.com/rust-lang/rust/pull/65694/files#diff-aee445fc8623471c31a62729dbf11e6eR79

I tried to resolve that by returning the terminators directly out of `find_eligible_blocks()` but then there's another issue caused by double borrows of `body` inside that function. I'm not sure how to resolve that.

Reply:
oh... how about mutably iterating over all blocks and instead of `find_eligible_blocks` giving you a `Vec` or an `Iterator` it just checks whether the block should be checked by returning `Option<Ty<'tcx>>::Some` or `None`. So instead of building up the list and then iterating over it, you iterate and skip all blocks that are not eligible

Reply:
I think that would work. I'll do that tonight.

Reply:
Fixed